### PR TITLE
Adding VirulenceFinder to TheiaProk

### PR DESCRIPTION
**Pull request by sage-wright**
This PR adds the VirulenceFinder module to all TheiaProk workflows.

## :hammer_and_wrench: Changes Being Made

- VirulenceFinder has been added to all TheiaProk workflows. 

## :brain: Context and Rationale

For now, VirulenceFinder will be run on only _E. coli_ samples. We are not certain if it is appropriate to run on various Shigella species. According to the [documentation](https://bitbucket.org/genomicepidemiology/virulencefinder/src/master/), _Enterococcus_, _S. aureus_, and _Listeria_ have databases as well, but the request for this tool was made specifically for _E. coli_. However, adding this module to apply to the other potential organisms would require only a relatively simple adjustment to the `merlin_magic` logic.

VirulenceFinder does not always run successfully on SE data, so in the case of SE samples (either Illumina and ONT data), the assembly will be used as input. For PE Illumina data, both reads will be used.

## :clipboard: Workflow/Task Steps 

1. Determine how to run VirulenceFinder based off of three Booleans: `paired_end`, `ont_data`, and `assembly_only`. If `paired_end` is true while `assembly_only` and `ont_data` are false, then VirulenceFinder will use both `read1` and `read2` as input. Otherwise, `assembly_fasta` will be used.
2. Run VirulenceFinder.
3. Parse the outputs. In the case where VirulenceFinder has found no hits, report that no hits were found instead of a blank string.

### Inputs

Mandatory:
- `samplename`, the three Boolean values listed above

Optional:
- `read1`, `read2`, `assembly_fasta` to enable different workflows to run the same task on different input data types

### Outputs

- `virulencefinder_report` which is the tab-delimited file containing all of the results
- `virulencefinder_docker` which is a string saying the docker image used
- `virulencefinder_factors` which is a comma-delimited string of the hits identified. Name subject to be changed.

### Locally

- [x] Local tests suceed as expected
Local test suceeding

### Terra

- [TheiaProk_Illumina_PE_PHB](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/6f18f284-8eed-4ddd-89cf-7dfd45f43b5f)
- [TheiaProk_Illumina_SE_PHB](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/405d8696-88a4-471c-9c87-f5f6924d5bdb)
outputs
- [TheiaProk_ONT_PHB](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/c7ddb43f-6d1d-4dbf-b9f7-26371bd2dadf) - the failure is an issue unrelated to this PR
- [TheiaProk_FASTA_PHB](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Wright_PHBG_Sandbox/job_history/078cd611-7036-4185-8ec6-f7e2d9fbf228)

## :microscope: Quality checks

Pull Request (PR) checklist:
- [x] Include a description of what is in this pull request in this message.
- [x] The workflow/task has been tested locally and on Terra
- [x] The CI/CD has been adjusted and tests are passing
- [x] Everything follows the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

### CODE REVIEW

**Comment by cimendes** [comment]
This PR look beautiful! A small suggestion, I would change the `virulencefinder_report` to `virulencefinder_report_tsv` to be in accordance with the style guide. I would change `virulencefinder_factors` to `virulencefinder_hits` or `virulencefinder_results` to be more intuitive on what are the contents of the column.

**Comment by sage-wright** [comment]
It helps someone made a beautiful template! 😜 I'll make those changes!

**Comment by cimendes** [comment]
@sage-wright this PR is marked as a draft. Is this on purpose? Is it okay to go ahead and test it for final review?

**Review by cimendes — APPROVED** [review]
Thanks for renaming the output columns, they seem much more intuitive to me now. :) 
image

I've tested the PR in 5 ecoli samples both in [PE ](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/80cc4354-c154-4d86-bf8b-9b946fea435a) and [SE ](https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/eef9555b-1376-44bb-9b21-8c1c5456be3a) (just dropped the second read) and it ran successfully on both datasets (yey!). 

I got different results (figures below) but this could be due to the different methodologies implemented (read based for paired-end, and assembly based for single-end). I do not have the source of truth for this dataset so I cannot make any comments regarding the accuracy of the tool. Functionality-wise I have no further comments. 

ecoli PE results:
image


ecoli SE results:
image

**Comment by kapsakcj** [comment]
I'm testing on Terra and running into a rare (only happening to 2 samples) and odd issue where virulencefinder is "stuck". Task has been running for 3+ hrs now...

Need to dive a little bit deeper to figure out why this is happening

https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/064a872d-089d-4651-b839-80b380e17bda

**Comment by kapsakcj** [comment]
Also, here's a sample ID where virulencefinder only found one hit `stx` when I expected it to have a long list of hits (based on amrfinderplus outputs as well as the virulencefinder webtool outputs

2009C-3606

**Comment by sage-wright** [comment]
@kapsakcj, I've added those optional parameters and exposed them to Terra

File: tasks/species_typing/task_virulencefinder.wdl [hunk 1]
```diff
@@ -49,6 +49,9 @@ task virulencefinder {
         -i ~{read1} ~{read2} \
         -o . \
         -tmp tmp \
+        ~{'-l ' + coverage_threshold} \
```
Comment (sage-wright) [inline tasks/species_typing/task_virulencefinder.wdl:52]: OMG thank you, I forgot I called the task twice 🤦

**Comment by kapsakcj** [comment]
I'm re-testing in Terra now that `-d virulence_ecoli` has been specified. Without that option, it uses all databases which includes non-Ecoli species databases

Want to review results for 2009C-3606 to ensure there is a bunch of hits found instead of only one, `stx`. 

https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/1f2bb768-0ec4-48be-b4d7-3b379f0f5940

**Comment by kapsakcj** [comment]
☹️ So I expected to see a lot of hits for 2009C-3606 and after the last commit, it returned nothing. Something is not right here...

**Comment by sage-wright** [comment]
Testing this in your sandbox, Curtis. Hope you don't mind. 
https://app.terra.bio/#workspaces/cdph-terrabio-taborda-manual/curtis_sandbox/job_history/ff656ddc-4072-43af-aeb5-674d1d70fbdc

**Comment by sage-wright** [comment]
That seemed to fix it. Something must be up with the reads.

```
Database	Virulence factor	Identity	Query / Template length	Contig	Position in contig	Protein function	Accession number
virulence_ecoli	AslA	94.57	1657 / 1656	contig00019 len=90846 cov=52.0 corr=0 origname=Contig_108_51.9896_pilon sw=shovill-se-skesa/1.1.0 date=20230601	51820..53475		CP054236
virulence_ecoli	anr	94.37	213 / 213	contig00052 len=9433 cov=33.0 corr=0 origname=Contig_119_32.9551_pilon sw=shovill-se-skesa/1.1.0 date=20230601	3967..4179	 AraC negative regulator	AL391753
virulence_ecoli	astA	91.3	115 / 117	contig00190 len=598 cov=45.5 corr=0 origname=Contig_160_45.4913_pilon sw=shovill-se-skesa/1.1.0 date=20230601	328..442	  Heat-stable enterotoxin EAST-1	AB042005
virulence_ecoli	astA	100.0	97 / 117	contig00063 len=6599 cov=52.4 corr=0 origname=Contig_3_52.4143_pilon sw=shovill-se-skesa/1.1.0 date=20230601	411..507	  Heat-stable enterotoxin EAST-1	HM099897
virulence_ecoli	chuA	100.0	1983 / 1983	contig00006 len=217340 cov=50.6 corr=0 origname=Contig_158_50.5729_pilon sw=shovill-se-skesa/1.1.0 date=20230601	16924..18906	  Outer membrane hemin receptor	UGGA01000003
virulence_ecoli	csgA	93.26	460 / 456	contig00020 len=87712 cov=52.2 corr=0 origname=Contig_253_52.167_pilon sw=shovill-se-skesa/1.1.0 date=20230601	79022..79480	 curlin major subunit CsgA	CP069646
virulence_ecoli	eae-g01-gamma	100.0	2805 / 2805	contig00013 len=145334 cov=50.3 corr=0 origname=Contig_56_50.3493_pilon sw=shovill-se-skesa/1.1.0 date=20230601	135797..138601		44
virulence_ecoli	ehxA	99.97	2997 / 2997	contig00038 len=29383 cov=27.6 corr=0 origname=Contig_250_27.5577_pilon sw=shovill-se-skesa/1.1.0 date=20230601	17085..20081	  Enterohaemolysin	AB011549
virulence_ecoli	espA	100.0	579 / 579	contig00013 len=145334 cov=50.3 corr=0 origname=Contig_56_50.3493_pilon sw=shovill-se-skesa/1.1.0 date=20230601	141342..141920	  Type III secretions system	AE005174
virulence_ecoli	espB	100.0	939 / 939	contig00013 len=145334 cov=50.3 corr=0 origname=Contig_56_50.3493_pilon sw=shovill-se-skesa/1.1.0 date=20230601	143078..144016	  Secreted protein B	AE005174
virulence_ecoli	espJ	100.0	654 / 654	contig00035 len=33129 cov=52.4 corr=0 origname=Contig_230_52.4348_pilon sw=shovill-se-skesa/1.1.0 date=20230601	799..1452	  Prophage-encoded type III secretion system effector	AE005174
virulence_ecoli	espP	99.97	3903 / 3903	contig00049 len=11300 cov=28.6 corr=0 origname=Contig_289_28.6328_pilon sw=shovill-se-skesa/1.1.0 date=20230601	136..4038	  Serine protease autotransporter EspP	AB011549
virulence_ecoli	espY2:000868321	100.0	570 / 570	contig00001 len=440574 cov=52.7 corr=0 origname=Contig_73_52.6537_pilon sw=shovill-se-skesa/1.1.0 date=20230601	145636..146205		
virulence_ecoli	etpD	100.0	1758 / 1758	contig00038 len=29383 cov=27.6 corr=0 origname=Contig_250_27.5577_pilon sw=shovill-se-skesa/1.1.0 date=20230601	4148..5905	  Type II secretion protein EtpD	AB011549
virulence_ecoli	fdeC	94.97	4254 / 4254	contig00004 len=268921 cov=52.4 corr=0 origname=Contig_237_52.4091_pilon sw=shovill-se-skesa/1.1.0 date=20230601	258540..262793	 intimin-like adhesin FdeC	AP010953
virulence_ecoli	fimH	100.0	489 / 489	contig00001 len=440574 cov=52.7 corr=0 origname=Contig_73_52.6537_pilon sw=shovill-se-skesa/1.1.0 date=20230601	321034..321522	   Type 1 fimbriae	NA
virulence_ecoli	gad	100.0	1178 / 1401	contig00017 len=96371 cov=52.6 corr=0 origname=Contig_192_52.5599_pilon sw=shovill-se-skesa/1.1.0 date=20230601	95194..96371	  Glutamate decarboxylase	BA000007
virulence_ecoli	gad	100.0	1178 / 1401	contig00026 len=66059 cov=50.8 corr=0 origname=Contig_241_50.7989_pilon sw=shovill-se-skesa/1.1.0 date=20230601	1..1178	  Glutamate decarboxylase	BA000007
virulence_ecoli	hlyE	98.47	918 / 918	contig00023 len=79166 cov=51.9 corr=0 origname=Contig_252_51.8719_pilon sw=shovill-se-skesa/1.1.0 date=20230601	71925..72842	  Avian E. coli haemolysin	ECU57430
virulence_ecoli	iha	100.0	2091 / 2091	contig00029 len=46442 cov=51.1 corr=0 origname=Contig_46_51.055_pilon sw=shovill-se-skesa/1.1.0 date=20230601	11436..13526	  Adherence protein	AE005174
virulence_ecoli	iss	97.96	294 / 294	contig00056 len=8118 cov=75.0 corr=0 origname=Contig_132_74.9953_pilon sw=shovill-se-skesa/1.1.0 date=20230601	2106..2399	  Increased serum survival	CP001509
virulence_ecoli	katP	100.0	2211 / 2211	contig00078 len=3700 cov=59.3 corr=0 origname=Contig_51_59.3139_pilon sw=shovill-se-skesa/1.1.0 date=20230601	844..3054	  Plasmid-encoded catalase peroxidase	AB011549
virulence_ecoli	nleA	100.0	1326 / 1326	contig00063 len=6599 cov=52.4 corr=0 origname=Contig_3_52.4143_pilon sw=shovill-se-skesa/1.1.0 date=20230601	4300..5625	  Non-LEE encoded effector A	CP001164
virulence_ecoli	nleB	100.0	981 / 981	contig00005 len=238205 cov=51.1 corr=0 origname=Contig_118_51.1428_pilon sw=shovill-se-skesa/1.1.0 date=20230601	237046..238026	  Non-LEE encoded effector B	AE005174
virulence_ecoli	nleB	100.0	990 / 990	contig00048 len=15107 cov=51.9 corr=0 origname=Contig_218_51.8951_pilon sw=shovill-se-skesa/1.1.0 date=20230601	9376..10365	  Non-LEE encoded effector B	AE005174
virulence_ecoli	nleC	100.0	993 / 993	contig00005 len=238205 cov=51.1 corr=0 origname=Contig_118_51.1428_pilon sw=shovill-se-skesa/1.1.0 date=20230601	235993..236985	  Non-LEE encoded effector C	AE005174
virulence_ecoli	nlpI	99.77	885 / 885	contig00003 len=280581 cov=52.3 corr=0 origname=Contig_123_52.3022_pilon sw=shovill-se-skesa/1.1.0 date=20230601	173731..174615	  lipoprotein NlpI precursor	CP000243
virulence_ecoli	ompT	100.0	954 / 954	contig00074 len=4085 cov=48.3 corr=0 origname=Contig_193_48.3114_pilon sw=shovill-se-skesa/1.1.0 date=20230601	1933..2886	  Outer membrane protease (protein protease 7)	AKKX01000238
virulence_ecoli	terC	100.0	1041 / 1041	contig00029 len=46442 cov=51.1 corr=0 origname=Contig_46_51.055_pilon sw=shovill-se-skesa/1.1.0 date=20230601	16120..17160	  Tellurium ion resistance protein	AE005174
virulence_ecoli	terC	98.74	714 / 714	contig00002 len=362209 cov=52.3 corr=2 origname=Contig_19_52.2873_pilon sw=shovill-se-skesa/1.1.0 date=20230601	201032..201745	  Tellurium ion resistance protein	CP007491
virulence_ecoli	terC	98.55	964 / 966	contig00003 len=280581 cov=52.3 corr=0 origname=Contig_123_52.3022_pilon sw=shovill-se-skesa/1.1.0 date=20230601	102721..103684	  Tellurium ion resistance protein	MG591698
virulence_ecoli	tir	100.0	1677 / 1677	contig00013 len=145334 cov=50.3 corr=0 origname=Contig_56_50.3493_pilon sw=shovill-se-skesa/1.1.0 date=20230601	133453..135129	  Translocated intimin receptor protein	AE005174
virulence_ecoli	toxB	99.95	9510 / 9510	contig00050 len=10443 cov=24.3 corr=0 origname=Contig_105_24.2551_pilon sw=shovill-se-skesa/1.1.0 date=20230601	83..9592	  Toxin B	AB011549
virulence_ecoli	traT	100.0	423 / 423	contig00029 len=46442 cov=51.1 corr=0 origname=Contig_46_51.055_pilon sw=shovill-se-skesa/1.1.0 date=20230601	45659..46081	  Outer membrane protein complement resistance	AKKX01000148
virulence_ecoli	traT	100.0	129 / 129	contig00029 len=46442 cov=51.1 corr=0 origname=Contig_46_51.055_pilon sw=shovill-se-skesa/1.1.0 date=20230601	35727..35855	  Outer membrane protein complement resistance	AMTE01000156
```

File: tasks/species_typing/task_virulencefinder.wdl [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    #     ~{'-l ' + coverage_threshold} \
+    #     ~{'-t ' + identity_threshold} \
+    #     ~{'-d' + database} \
+    #     -x
+    # fi
+   
+    # rename output file
+    mv results_tab.tsv ~{samplename}_results_tab.tsv
+
+    # parse 2nd column (Virulence factor) of _tab.tsv file into a comma-separated string
```
Comment (kevinlibuit) [inline tasks/species_typing/task_virulencefinder.wdl:63]: Any value in keeping this commented block?

**Review by kevinlibuit — APPROVED** [review]

**Pull request merged**